lang/ghc/Portfile

```diff
@@ -29,144 +24,173 @@ long_description    \
-if { ${os.platform} eq "darwin" && ( ${os.major} < 13 || ${os.major} == 15 ) } {
-    compiler.whitelist    macports-clang-4.0
-    depends_run-append    port:clang-4.0
+build {
+    ui_msg "\tBuild performed in destroot."
 }
 
-set bootstraproot ${prefix}/share/ghc-bootstrap
-set llvmPrefix  ${prefix}/libexec/llvm-3.5
-configure.args  --with-ghc=${bootstraproot}/bin/ghc \
-                --with-iconv-includes=${prefix}/include \
-                --with-iconv-libraries=${prefix}/lib \
-                --with-gmp-includes=${prefix}/include \
-                --with-gmp-libraries=${prefix}/lib \
-                --with-ffi-includes=${prefix}/lib/libffi-3.2.1/include \
-                --with-ffi-libraries=${prefix}/lib \
-                --with-system-libffi \
-                --with-gcc="${configure.cc}" \
-                --with-llc=${llvmPrefix}/bin/llc \
-                --with-opt=${llvmPrefix}/bin/opt
-
-# OK so because the bootstrap binary has been prebuilt for libraries in
-# /usr/lib we search these in addition to MacPorts' paths to prevent link
-# errors. GHC _should_ actually compile itself in stage2 using paths from the
-# command line arguments
-compiler.cpath /usr/include
-compiler.library_path /usr/lib
-
-build.args      VERBOSE=1
-destroot.args   VERBOSE=1
-
-pre-activate {
-    set obsoletes [list]
-
-    # Legacy port deactivation hack added 2012-12-08, hs-platform-ghc no longer exists
-    lappend obsoletes hs-platform-ghc
-    # Legacy port deactivation hack added 2013-08-11, hs-process is provided by base
-    lappend obsoletes hs-process
-    # Legacy port deactivation hack added 2013-08-14, hs-template-haskell is provided by base
-    lappend obsoletes hs-template-haskell
-    # Legacy port deactivation hack added 2015-01-05, hs-transformers, hs-xhtml
-    # and hs-binary are provided by base
-    lappend obsoletes hs-transformers
-    lappend obsoletes hs-xhtml
-    lappend obsoletes hs-binary
-
-    foreach obsolete $obsoletes {
-        if {![catch {set obsolete_installed [lindex [registry_active $obsolete] 0]}]} {
-            # $obsolete is installed and active
-            # force deactivation
-            registry_deactivate_composite $obsolete "" [list ports_nodepcheck 1]
-        }
+destroot {
+    ui_msg "\tBuild/destroot from ${name}-${version}-x86_64-apple-darwin${extract.suffix}…"
+    system -W ${worksrcpath} "make install -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
+
+    if {[variant_isset "bootstrap"]} {
+        ui_msg "\tBootstrap build/destroot from ${name}-${version}-src${extract.suffix}…"
+        xinstall -m 0755 -d ${workpath}/src
+        system -W ${workpath}/src \
+            "tar xJf ${distpath}/${name}-${version}-src${extract.suffix}"
+        system -W ${workpath}/src/${name}-${version} \
+            "./boot"
+        system -W ${workpath}/src/${name}-${version} \
+            "export PATH=\${PATH}:${destroot}${prefix}/bin ; \
+                ./configure --prefix=${destroot}${prefix}"
```

Comment:
All these steps use a lot of functionality provided by regular port actions (extract, configure etc.). It might be better to create a subport for bootstrapping purposes that would perform all those steps in the regular phases (extracting, configuring, ...) and then the main port could add a build dependency on that bootstrapping subport.

Reply:
I’m allergic to an overly complicated build approach, which is main reason this port is years out of date—extra `-bootstrap` port, `hs-*` ports, etc.

Also, I don’t know how subports work. If you feel strongly about this, would you please sketch out the tcl code and file layout?

Because bootstrapping doesn’t even work yet, I’m inclined to leave the current approach as is.

Reply:
See discussion below. I'm not aware of a way around this issue.